### PR TITLE
Vnext/491 update database on start

### PR DESCRIPTION
**Pull request by gregglarson**
With all cash flow fixes.

### CODE REVIEW

File: iAM-legacy/Simulation/AppliedTreatment.cs [hunk 1]
```diff
@@ -24,6 +24,10 @@ public class AppliedTreatment
         String m_strChangeHash;
         AppliedTreatment _multipleTreatment;//Stores a follow up treatment.
         bool _isExclusive;
+        //For split treatment (before split).
```
Comment (ARA-Trans) [inline iAM-legacy/Simulation/AppliedTreatment.cs]: The `AppliedTreatments` class is violating encapsulation by making its properties publicly settable. The class needs to be changed to have the properties passed as parameters to the constructor and all of the public `set`'s should be made private.

File: iAM-legacy/Simulation/Committed.cs [hunk 2]
```diff
@@ -30,6 +30,10 @@ public class Committed
         int _resultType;
         //If this is set. The committed project is a ScheduledTreatment and not a true committed project.
         public string ScheduledTreatmentId { get; set; }
+        public string SplitTreatmentId { get; set; }
```
Comment (ARA-Trans) [inline iAM-legacy/Simulation/Committed.cs]: These properties should also be made private and have the committed project injected into the appropriate constructors or accessed via ID.

File: iAM-legacy/Simulation/Investments.cs [hunk 3]
```diff
@@ -148,11 +152,16 @@ public void LoadBudgets()
             String strYear;
 
             String strAmount;
-
+            MaximumYear = Int32.MinValue;
```
Comment (ARA-Trans) [inline iAM-legacy/Simulation/Investments.cs]: `MaximumYear` is only used in the investment class. Make it private.

File: iAM-legacy/Simulation/Investments.cs [hunk 4]
```diff
@@ -211,14 +232,44 @@ public void LoadBudgets()
...
+        public string IsBudgetAvailable(float fAmount, String strBudget, String strYear, Hashtable hashAttributeValue,Priorities priority, List<ISplitTreatmentLimit> limits, string budgetLimitType, out string budgetHash, out ISplitTreatmentLimit splitTreatmentLimit)
         {
+            
+            splitTreatmentLimit = limits[0];
+
+            //Check priority.  If not possible just return null.
+            if (!(priority.IsAllSections || priority.Criteria.IsCriteriaMet(hashAttributeValue)))
+            {
+                budgetHash = "";
+                return null;
```
Comment (ARA-Trans) [inline iAM-legacy/Simulation/Investments.cs]: Avoid returning null in API calls.

File: iAM-legacy/Simulation/Investments.cs [hunk 5]
```diff
@@ -321,104 +353,96 @@ public string IsBudgetAvailable(float fAmount, String strBudget, String strYear,
...
+                    catch (Exception e)
                     {
-                        if (priority.Criteria.IsCriteriaMet(hashAttributeValue))
-                        {
-                            float fPercentLimit = (float)priority.BudgetPercent[budgetCheck];
-                            var original = (float)priority.BudgetPercent[budgetCheck] * fOriginal;
-                            budgetHash = fAmount + "/" + fAvailable.ToString("f0") + "(" + original.ToString("f2") + ")";
-                            if (fPercent < fPercentLimit) return budgetCheck;
-                        }
+                        SimulationMessaging.AddMessage(new SimulationMessage("Error: Evaluating priority. " + e.Message));
```
Comment (ARA-Trans) [inline iAM-legacy/Simulation/Investments.cs]: `SimulationMessaging` is unusable in the web app. We need to have this written out to an error log file that we can get from PennDOT IT.

File: iAM-legacy/Simulation/Sections.cs [hunk 6]
```diff
@@ -220,6 +220,9 @@ public int RollToYear
             set { m_nStartYear = value-1; }
         }
 
+        public Dictionary<int,float> ScheduledCost { get; set; }
```
Comment (ARA-Trans) [inline iAM-legacy/Simulation/Sections.cs]: This needs to be made private

File: iAM-legacy/Simulation/SplitTreatment.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        public string Id { get; }
+
+        public string Description { get; }
+
+        public Criterias Criteria { get; }
+
+        public List<ISplitTreatmentLimit> Limits { get; }
+
+
+        public SplitTreatment(string splitTreamentId, string description, string criteria,  List<ISplitTreatmentLimit> limits)
```
Comment (ARA-Trans) [inline iAM-legacy/Simulation/SplitTreatment.cs]: Good injection of properties here.

File: iAM-legacy/Simulation/SplitTreatment.cs [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        {
+            Id = splitTreamentId;
+            Description = description;
+            Limits = limits;
+            Criteria = new Criterias("SPLIT_TREATMENT", "BINARY_CRITERIA", splitTreamentId);
+            byte[] assemblyCriteria = null;
+            string currentCriteria = "";
+            if (!string.IsNullOrWhiteSpace(criteria)) currentCriteria = Simulation.ConvertOMSAttribute(criteria);
```
Comment (ARA-Trans) [inline iAM-legacy/Simulation/SplitTreatment.cs]: There appears to be a lot going on in the constructor. I'd consider making some private functions to operate on the data instead, or introduce a new set of classes to do that work and inject them in.

File: iAM-legacy/Simulation/SplitTreatmentLimit.cs [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        public int Rank { get; }
+
+        public float Amount { get; }
+
+        public List<float> Percentages { get; }
+
+
+        public SplitTreatmentLimit(string rank, string amount, string percentages)
+        {
+            Rank = Convert.ToInt32(rank);
```
Comment (ARA-Trans) [inline iAM-legacy/Simulation/SplitTreatmentLimit.cs]: Same here, consider making some private functions to handle the data operations or preferably some new classes.

File: iAM-legacy/Simulation/Committed.cs (hunk 2 above)
Comment (gregglarson) [reply iAM-legacy/Simulation/Committed.cs]: Added constructor to committed for split treatment

File: iAM-legacy/Simulation/Investments.cs (hunk 3 above)
Comment (gregglarson) [reply iAM-legacy/Simulation/Investments.cs]: Made private

File: iAM-legacy/Simulation/Investments.cs (hunk 4 above)
Comment (gregglarson) [reply iAM-legacy/Simulation/Investments.cs]: Returns empty string.  The consumer of this function was asking for a budget that had money and met criteria.  Returning null was showing that no budget was available.  Return value was being checked IsStringNullEmptyWhitespace

File: iAM-legacy/Simulation/Investments.cs (hunk 5 above)
Comment (gregglarson) [reply iAM-legacy/Simulation/Investments.cs]: Not a useful message for the web application.  More for debugging on the desktop.  Will properly wrap messages for the web.

File: iAM-legacy/Simulation/Sections.cs (hunk 6 above)
Comment (gregglarson) [reply iAM-legacy/Simulation/Sections.cs]: Not used anymore. Removed

File: iAM-legacy/Simulation/SplitTreatment.cs (hunk 8 above)
Comment (gregglarson) [reply iAM-legacy/Simulation/SplitTreatment.cs]: Agree, but follows the pattern of about 20 other classes that initialize criteria.  Will refactor all in future.

File: iAM-legacy/Simulation/Investments.cs (hunk 5 above)
Comment (ARA-Trans) [reply iAM-legacy/Simulation/Investments.cs]: Change the message to be a useful log message and wrap in the WebAPI logging call.

**Pull request merged**